diffraction/WISH/bragg-detect/cnn/WISHMLBraggPeaksDetector.py

```diff
@@ -0,0 +1,108 @@
+        for ipk, pk in enumerate(peaksws):
+            pk.setIntensity(filtered_indices[ipk, -1])
+
+        #Filter duplicates by qlab
+        BaseSX.remove_duplicate_peaks_by_qlab(peaksws, q_tol)
+        print(f"Bragg peaks finding from FasterRCNN model is completed in {time.time()-start_time} seconds!")
+
+
+    def _do_cnn_inferencing(self, ws_name):
+        print(f"Do CNN inferencing for {ws_name}")
+        data_set = WISHWorkspaceDataSet(ws_name)
+        data_loader = tc.utils.data.DataLoader(data_set, batch_size=self.batch_size, shuffle=False, num_workers=self.workers)
+        self.model.eval()
+        predicted_indices_with_score = []
+        for batch_idx, img_batch in enumerate(tqdm(data_loader, desc="Processing batches of tubes")):
+            for img_idx, img in enumerate(img_batch):
+                tube_idx = batch_idx * data_loader.batch_size + img_idx
+                with tc.no_grad():
+                    prediction = self.model([img.to(self.device)])[0]
+                    nms_prediction = self._apply_nms(prediction, self.iou_threshold)
+                    for box, score in zip(nms_prediction['boxes'], nms_prediction['scores']):
+                        tof = (box[0]+box[2])/2
+                        tube_res = (box[1]+box[3])/2
+                        predicted_indices_with_score.append([tube_idx, tube_res.cpu(), tof.cpu(), score.cpu()])
+        return data_set, np.array(predicted_indices_with_score)
+                    
+
+    def _apply_nms(self, orig_prediction, iou_thresh):
+        keep = torchvision.ops.nms(orig_prediction['boxes'], orig_prediction['scores'], iou_thresh)
+        final_prediction = orig_prediction
+        final_prediction['boxes'] = final_prediction['boxes'][keep]
+        final_prediction['scores'] = final_prediction['scores'][keep]
+        final_prediction['labels'] = final_prediction['labels'][keep]
+        return final_prediction
+
+
+    def _select_device(self):
+        if tc.cuda.is_available():
+            print("GPU device is found!")
+            return tc.device("cuda")
+        else:
+            warnings.warn(
+                "Warning! GPU is not available, the program will run very slow..", RuntimeWarning)
+            return tc.device("cpu")
+
+
+    def _load_cnn_model_from_weights(self, weights_path):
+        model = self._get_fasterrcnn_resnet50_fpn(num_classes=2)
+        model.load_state_dict(tc.load(weights_path, map_location=self.device))
+        return model.to(self.device)
+
+
+    def _get_fasterrcnn_resnet50_fpn(self, num_classes=2):
+        model = torchvision.models.detection.fasterrcnn_resnet50_fpn(weights=None)
+        in_features = model.roi_heads.box_predictor.cls_score.in_features
+        # replace the head
+        model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes) 
+        return model
+    
+if __name__ == "__main__" or __name__ == "mantidqt.widgets.codeeditor.execution":
```

Comment:
Do you still need this execution code - perhaps you could add this example to the doc string of the class?